File: trade_analyzer.rb
require 'JSON'
FILE = 'trades.txt'

class Position
	attr_reader :q_id, :q_name, :c_name, :assumption_q_id, :assumption_q_name, :assumption_c_name, :exposure
	def initialize(args)
		@q_id = args[:q_id]
		@q_name = args[:q_name]
		@c_name = args[:c_name]
		@assumption_q_id = args[:assumption_q_id]
		@assumption_q_name = args[:assumption_q_name]
		@assumption_c_name = args[:assumption_c_name]
		@exposure = args[:exposure]
		@start_value = args[:start_value]
		@end_value = args[:end_value]
	end

	def to_tsv
		"#{@q_id}\t#{@q_name}\t#{@c_name}\t#{@assumption_q_id}\t#{@assumption_q_name}\t#{@assumption_c_name}\t#{@start_value}\t#{@end_value}\t#{@exposure}"
	end
end

class ScicastPosition
	@@position = []
	@@edits = []
	@@username = ""
	def self.parse_edits
		file = File.open(FILE)
		file_text = ""
		file.each do |line|
			file_text << line
		end
		raw_edits = JSON.parse(file_text)
		raw_edits.each do |edit|
			edit["question"]["choices"].each_with_index do |choice, index|
				exposure = edit["assets_per_option"][index].round(3)
				if exposure != 0 && edit["question"]["settled_values"] == nil
				 	args = {q_id: edit["question"]["id"], 
				 					q_name: edit["question"]["name"].gsub(/[\n\t\r]/,""), 
				 					c_name: choice["name"].gsub(/[\n\t\r]/,""),
				 					exposure: exposure,
				 					assumption_q_id: "#{edit["assumptions"][0]["id"] unless edit["assumptions"] == []}",
				 					assumption_q_name: "#{edit["assumptions"][0]["name"].gsub(/[\n\t\r]/,"") unless edit["assumptions"] == []}",
				 					assumption_c_name: "#{edit["assumptions"][0]["choices"][edit["assumptions"][0]["dimension"]]["name"].gsub(/[\n\t\r]/,"") unless edit["assumptions"] == []}",
				 					start_value: edit["old_value_list"][index].round(5),
				 					end_value: edit["new_value_list"][index].round(5)}
				 	@@edits << Position.new(args) 
				end
			end
		end
		# p raw_edits.last["old_value_list"]
		# p raw_edits.last["new_value_list"]
		@@username = raw_edits[0]["user"]["username"]
		@@date = Time.now.to_s[0..9]
	end

	def self.aggregate_position
		@@edits.group_by{|edit| [edit.q_id, edit.q_name, edit.c_name, edit.assumption_q_id, edit.assumption_q_name, edit.assumption_c_name]}.each do |k,v|
			exposure = v.map{|edit| edit.exposure}.inject(:+).round(3)
			@@position << k + [exposure]
		end
	end

	def self.save_position_as_tsv
		IO.write("#{@@username}-position-#{@@date}.tsv",self.position_to_tsv)
	end

	def self.save_edits_as_tsv
		IO.write("#{@@username}-edits-#{@@date}.tsv",self.edits_to_tsv)
	end

	private 

	def self.edits_to_tsv
		output = "Question id\tQuestion name\tChoice name\tAssumption q id\tAssumption q name\tAssumption choice name\tStart value\tEnd value\tExposure\n"
		@@edits.each {|edit| output << edit.to_tsv + "\n"}
		output
	end

	def self.position_to_tsv
		output = "Question id\tQuestion name\tChoice name\tAssumption q id\tAssumption q name\tAssumption choice name\texposure\n"
		@@position.each {|position| output << position.join("\t") + "\n"}
		output
	end
end

ScicastPosition.parse_edits
ScicastPosition.aggregate_position
ScicastPosition.save_edits_as_tsv
ScicastPosition.save_position_as_tsv

File: trade_analyzer.py
#!/usr/bin/python

import json
from operator import attrgetter
from itertools import groupby
import urllib2

URL = "https://scicast.org/trades/?user_id={0}&include_current_probs=False"


FILENAME = 'trades.txt'
ALL_KEYS = ['q_id', 'q_name', 'c_name', 'assumption_q_id', 'assumption_q_name',
            'assumption_c_name', 'exposure']
KEYS     = ALL_KEYS[:-1]

def download_trades(uid=1, file_name=FILENAME):
    url = URL.format(uid)
    u = urllib2.urlopen(url)
    f = open(file_name, 'wb')
    meta = u.info()
    file_size = int(meta.getheaders("Content-Length")[0])
    print "Downloading: %s Size: %s MiB" % (file_name, file_size // 1e6)
    
    file_size_dl = 0
    block_sz = 8192
    while True:
        buffer = u.read(block_sz)
        if not buffer:
            break
            
        file_size_dl += len(buffer)
        f.write(buffer)
        status = r"%10d  [%3.2f%%]" % (file_size_dl, file_size_dl * 100. / file_size)
        status = status + chr(8)*(len(status)+1)
        print status,
    f.close()


class Position(object):
    def __init__(self, **data):
        for key in ALL_KEYS:
            setattr(self, key, data[key])

    def to_tsv(self):
        fields = [getattr(self, key) for key in ALL_KEYS]
        return "\t".join(['{0}'.format(field) for field in fields])


def _remove_whitespace(string_):
    for char in ['\n', '\t', '\r']:
        string_ = string_.replace(char, '')
    return string_

class ScicastPosition(object):
    def __init__(self, uid=None):
        self.position = []
        self.edits = []
        if uid is not None:
            download_trades(uid=uid)

    def parse_edits(self, filename=FILENAME, exclude_resolved=True):
        with open(filename, 'r') as f:
            data = json.loads(f.readlines()[0])
        for edit in data:
            if exclude_resolved and edit['asset_resolution'] is not None:
                continue
            for ix, choice in enumerate(edit['question']['choices']):
                edit_data = {
                    'q_id'              : edit['question']['id'],
                    'q_name'            : _remove_whitespace(
                        edit['question']['name']
                    ),
                    'c_name'            : _remove_whitespace(
                        choice['name']
                    ),
                    'exposure'          : edit['assets_per_option'][ix],
                    'assumption_q_id'   : (None if edit['assumptions'] == [] else
                                         edit['assumptions'][0]['id']),
                    'assumption_q_name' : (None if edit['assumptions'] == [] else
                                           _remove_whitespace(
                                               edit['assumptions'][0]['name']
                                           )
                                       ),
                    'assumption_c_name' : (None if edit['assumptions'] == [] else
                                           _remove_whitespace(
                                               edit['assumptions'][0]['choices'][
                                                   edit['assumptions'][0]['dimension']
                                               ]['name']
                                           )
                                       )
                }
                self.edits.append(Position(**edit_data))
                self.edits.sort(key=attrgetter(*KEYS))

    def aggregate_position(self):
        for (q_id, q_name, c_name,
             assumption_q_id,
             assumption_q_name,
             assumption_c_name), edits in groupby(self.edits, key=attrgetter(*KEYS)):
            exposure = sum(getattr(edit, 'exposure') for edit in edits)
            position_data = dict([(key, locals()[key]) for key in ALL_KEYS])
            self.position.append(Position(**position_data))

    def save_position_as_tsv(self):
        with open('position.tsv', 'w') as f:
            f.write(self._position_to_tsv())
    def save_edits_as_tsv(self):
        with open('edits.tsv', 'w') as f:
            f.write(self._edits_to_tsv())

    def _position_to_tsv(self):
        output = ("Question id\tQuestion name\tChoice name\t"
                  "Assumption q id\tAssumption q name\t"
                  "Assumption choice name\texposure\n")
        for position in self.position:
            output += '\t'.join([repr(getattr(position, key)) for key in ALL_KEYS])
            output += '\n'
        return output

    def _edits_to_tsv(self):
        output = ("Question id\tQuestion name\tChoice name\t"
                  "Assumption q id\tAssumption q name\t"
                  "Assumption choice name\texposure\n")
        for position in self.edits:
            output += '\t'.join([repr(getattr(position, key)) for key in ALL_KEYS])
            output += '\n'
        return output


def go(exclude_resolved=True):
    parser = ScicastPosition()
    print 'Parsing...'
    parser.parse_edits(exclude_resolved=exclude_resolved)
    print 'done'
    print 'Aggregating ...'
    parser.aggregate_position()
    print 'done'
    print 'Saving...'
    parser.save_edits_as_tsv()
    parser.save_position_as_tsv()
    print 'done'

                
go()
